src/pages/iou/MoneyRequestModal.js

```diff
@@ -0,0 +1,446 @@
+    }),
+
+    ...withLocalizePropTypes,
+};
+
+const defaultProps = {
+    hasMultipleParticipants: false,
+    report: {
+        participants: [],
+    },
+    iouType: CONST.IOU.IOU_TYPE.REQUEST,
+    currentUserPersonalDetails: {
+        localCurrencyCode: CONST.CURRENCY.USD,
+    },
+    personalDetails: {},
+    iou: {
+        creatingIOUTransaction: false,
+        error: false,
+        selectedCurrencyCode: null,
+    },
+};
+
+// Determines type of step to display within Modal, value provides the title for that page.
+const Steps = {
+    IOUAmount: 'iou.amount',
+    IOUParticipants: 'iou.participants',
+    IOUConfirm: 'iou.confirm',
+};
+
+const MoneyRequestModal = (props) => {
+    const reportParticipants = lodashGet(props, 'report.participants', []);
+    const participantsWithDetails = _.map(OptionsListUtils.getPersonalDetailsForLogins(reportParticipants, props.personalDetails), personalDetails => ({
+        login: personalDetails.login,
+        text: personalDetails.displayName,
+        firstName: lodashGet(personalDetails, 'firstName', ''),
+        lastName: lodashGet(personalDetails, 'lastName', ''),
+        alternateText: Str.isSMSLogin(personalDetails.login) ? Str.removeSMSDomain(personalDetails.login) : personalDetails.login,
+        icons: [{
+            source: ReportUtils.getAvatar(personalDetails.avatar, personalDetails.login),
+            name: personalDetails.login,
+            type: CONST.ICON_TYPE_AVATAR,
+        }],
+        keyForList: personalDetails.login,
+        payPalMeAddress: lodashGet(personalDetails, 'payPalMeAddress', ''),
+        phoneNumber: lodashGet(personalDetails, 'phoneNumber', ''),
+    }));
+
+    // Skip IOUParticipants step if participants are passed in
+    const steps = reportParticipants.length ? [Steps.IOUAmount, Steps.IOUConfirm] : [Steps.IOUAmount, Steps.IOUParticipants, Steps.IOUConfirm];
+
+    const prevCreatingIOUTransactionStatusRef = useRef(lodashGet(props.iou, 'creatingIOUTransaction'));
+    const prevNetworkStatusRef = useRef(props.network.isOffline);
+
+    const [previousStepIndex, setPreviousStepIndex] = useState(0);
+    const [currentStepIndex, setCurrentStepIndex] = useState(0);
+    const [participants, setParticipants] = useState(participantsWithDetails);
+    const [amount, setAmount] = useState('');
+    const [comment, setComment] = useState('');
+
+    useEffect(() => {
+        PersonalDetails.openMoneyRequestModalPage();
+        IOU.setIOUSelectedCurrency(props.currentUserPersonalDetails.localCurrencyCode);
+    // eslint-disable-next-line react-hooks/exhaustive-deps -- props.currentUserPersonalDetails will always exist from Onyx
+    }, []);
+
+    useEffect(() => {
+        // We only want to check if we just finished creating an IOU transaction
+        // We check it within this effect because we're sending the request optimistically but if an error occurs from the API, we will update the iou state with the error later
```

Comment:
Is it kind of sus and/or confusing? If the error occurs from the API then could we just look at `props.iou.error` as a dependency? If we didn't have an error before, but have one now then we should set the index to 0.

also having trouble understanding the case where we'd dismiss the modal. Anytime the `iou` changes and we have no errors and we're not in the process of creating the transaction via the API we call `dismissModal()`?

Reply:
Yeah agreed this doesn't make much sense in the hooks world. Going to make some changes to this

Reply:
Yeah I agree it is kind of confusing

> If the error occurs from the API then could we just look at props.iou.error as a dependency?

We could, but we'd still need to check the previous props to see if we are done creating iou otherwise this would run each re-render

> also having trouble understanding the case where we'd dismiss the modal. Anytime the iou changes and we have no errors and we're not in the process of creating the transaction via the API we call dismissModal()? Why is that exactly?

Don't have a great answer for this, I'm mostly just refactoring the existing logic in place

Reply:
My guess actually is that we have this logic in place for the offline handling of creating IOUs. Like if you dismiss the modal right after the transaction was created, then we don't have the ability to re-open and restart the money request. This handling is basically like - user comes back online? okay open the money request modal and then now that they are back online based on the success state of the IOU either restart the modal or dismiss it

Reply:
Ok, I think at a minimum we should just add a more specific dependency for `creatingIOUTransaction`? If I am understanding the logic correctly we only want to do "something" (show error or dismiss) when the transaction API call completes.

Reply:
I believe we're already kind of handling this with prevCreatingIOUTransactionStatusRef, like we need to check that creatingIOUTransaction is true but also that the previous creatingIOUTransaction is false to confirm this is the first time that the IOU was created

Reply:
More weirdness happening, creating this issue https://github.com/Expensify/App/issues/16942 to investigate it more deeply but not block on this